### ISSUE TITLE
[bug] `-> _` inference fails for `&'static str`

### ISSUE THREAD

**Issue by dev-ardi** [rust-lang/rust-analyzer#16676]
**rust-analyzer version**: 2023-12-18  

**rustc version**: 1.76
```rs
fn infer() -> _ {
    "hi"
}
```
![image](https://github.com/rust-lang/rust-analyzer/assets/47633543/b053b80b-e7a1-4ef6-b4af-3faa67b61b89)
Has the correct code action (3) but if we choose 2 we get

```rs
fn infer() -> &str {
    "hi"
}
```
which doesn't compile.

**Comment by Young-Flash** [comment]
related to [add_return_type](https://github.com/rust-lang/rust-analyzer/blob/5346002d07d09badaf37949bec68012d963d61fc/crates/ide-assists/src/handlers/add_return_type.rs) assist

**Comment by lnicola** [comment]
"replace with the correct return type: `S<'static>`" is from `cargo check`.

**Comment by dev-ardi** [comment]
Is that a rustc "error" (?) then?

**Comment by lnicola** [comment]
Try `cargo check` on that code and you'll see what I mean.

**Comment by dev-ardi** [comment]
Yes, I know what you mean but the fact that there is a difference between `cargo check` and r-a should be considered a bug, right?

**Comment by lnicola** [comment]
Maybe. The rust-analyzer diagnostics and assists are completely separate from the ones in the compiler. Most of the time they're worse, but I've seen cases where the RA fixes were better than the compiler's.
